designs/credential-update.rst

```diff
@@ -0,0 +1,154 @@
+
+Credential Update and Onboarding Workflow
+-----------------------------------------
+
+Letting users update their own credentials, but also allowing new users to create their own credentials
+needs a new workflow. Since these are nearly identical, these processes can be combined.
+
+Major considerations are:
+
+* Ability to start the credential creation (onboarding) work flow from a secured link.
+* Delegation of this permission from accounts to provide a reset path to users when needed.
+* Ensure that any credential updates are consistent and atomic.
+* Improved client side feedback around credential policy and rules.
+
+Initiation of the Credential Update Process
+===========================================
+
+The start of this process is that a user requests a credential update to begin for
+themself, or on behalf of another user.
+
+Self Update Workflow
+^^^^^^^^^^^^^^^^^^^^
+
+The user signals intent to update their credentials.
+
+An ACP check is made to check that the user has self-write to the relevant credential fields. If they
+do not, the credential update session is rejected.
+
+If they do have the access, a new credential update session is created, and the user is given a time
+limited token allowing them to interact with the credential update session.
+
+If the credental update session is abandoned it can not be re-accessed until it is expired.
+
+Onboarding/Reset Account Workflow
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The user signals intent that another users credentials should be updated.
+
+An ACP check is made to assert that the user has rights over the target users credential fields.
+If they do not, the update session is rejected. Note that the target user does *NOT* need self
+write permissions in this situation, which allows for a constrained permission on the target
+users (IE anonymous has no self write).
+
+If the access exists, a intent token is created into a link which can be provided to the user.
+
+Exchange of this intent token, creates the time limited credential update session token.
+
+This allows the intent token to have a seperate time window, to the credential update session token.
+
+If the intent token creates a credential update session, and the credential update session is *not*
+commited, it can be re-started by the intent token.
+
+If the credential update session has been committed, then the intent token can NOT create new
+credential update sessions (it is once-use).
+
+Consistency
+^^^^^^^^^^^
+
+Only one credential update process may exist at a time for a user. Attempts to create a second
+credential update session while an existing session exists is rejected until the previous session
+is committed or timed out.
+
+The credential update session can be rejected by the user, canceling it's progress.
+
+A credential update session IS tied to a single server, similar to authentication.
```

Comment:
This feels like a footgun, unless it'll be solved in the future - it'll cause issues for two cases I can think of:

 - load balanced IDM servers, or 
 - a session being created by an "internal" admin, for a user who's on the "public" network

Reply:
these two are interesting cases.

> load balanced IDM servers, or

This is already an issue with authentication, because we need to prevent classes of replay attacks and such. The auth-cred-change is a bit less prone to that attack flow, but for consistency it still makes more sense to keep this locked to a single state. We can't use / rely on backend sync due to eventual consistency. So we currently do already expect a load balancer to perform sticky sessions, which honestly is pretty widely required. Even LDAP requires it .... so for now, I'm happy to leave us requiring single-server contact and sticky sessions, and if it becomes an issue later, we should confront it at that time. 

> a session being created by an "internal" admin, for a user who's on the "public" network

This is already (mostly) solved! In the design the admin creates the intent-token. The intent token is currently "stateless" and then when that's given to the user, that intent token is exchanged for the credential update session. Because the user was the one that exchanged the intent -> credupdatesession that's done on the public network, so the credupdatesession will be on the public side and the user can proceed without issue. 

The main issue there is "what happens if the intent token is used multiple times?" Which is something I need to think about and factor into the design. I would like it to be that it can be exchanged multiple times for credupdatessessions, but only until *at least* one update session succeeds. After that one succeeds, it's now 'invalid'. So I have to think about how to achieve that nicely. It's not a super big deal, but consider we are on boarding someone, that's basically a 24 hour "change your credentials" link.

Reply:
Updated docs to address the intent token case.